qcore/shared.py

```diff
@@ -1,109 +1,204 @@
+            comment="#",
+            sep=r"\s+",
+            header=None,
+            names=["longitude", "latitude", "station"],
+            skipinitialspace=True,
+        )
+
     if not locations:
-        return stations
-    return (stations, station_lats, station_lons)
+        return stations["station"].to_list()
+    return (
+        stations["station"].to_list(),
+        stations["longitude"].to_list(),
+        stations["latitude"].to_list(),
+    )
 
 
-def get_corners(model_params, gmt_format=False):
+def get_corners(
+    model_params_ffp: Union[Path, str], gmt_format: bool = False
+) -> Union[list[tuple[float, float]], tuple[list[tuple[float, float]], str]]:
     """
     Retrieve corners of simulation domain from model params file.
-    model_params: file path to model params
-    gmt_format: if True, also returns corners in GMT string format
+
+    Parameters
+    ----------
+    model_params : Path or str
+        The file path of the model_params file.
+    gmt_format : bool, default False
+        If True, also returns corners in GMT string format.
+
+    Returns
+    -------
+    list[tuple[float, float]]
+        A list of (lon, lat) pairs read from the model params file.
+    str if gmt_format is True
+        A gmt format string specifying the corners read from the model params file
     """
     # with -45 degree rotation:
     #   c2
     # c1  c3
     #   c4
-    corners = []
-    with open(model_params, "r") as vmpf:
-        lines = vmpf.readlines()
-        # make sure they are read in the correct order at efficiency cost
-        for corner in ["c1=", "c2=", "c3=", "c4="]:
-            for line in lines:
-                if corner in line:
-                    corners.append(list(map(float, line.split()[1:3])))
-                    break
+    with open(model_params_ffp, "r") as model_params_file_handle:
+        corners = {}
+        for line in model_params_file_handle:
+            corner_match = re.match(
+                r"\s+c(\d)=\s+([0-9\.\-\+]+)\s+([0-9\.\-\+]+)", line
+            )
+            if corner_match:
+                corners[int(corner_match.group(1)) - 1] = (
+                    float(corner_match.group(2)),
+                    float(corner_match.group(3)),
+                )
+        corners = [corners[i] for i in range(4)]
+
     if not gmt_format:
         return corners
     # corners in GMT format
-    cnr_str = "\n".join([" ".join(map(str, cnr)) for cnr in corners])
+    cnr_str = "\n".join(f"{lon} {lat}" for lon, lat in corners)
     return corners, cnr_str
 
 
-def non_blocking_exe(cmd, debug=True, shell=False, stdout=True, stderr=True, **kwargs):
+def non_blocking_exe(
+    cmd: Union[str, list[str]],
+    debug: bool = True,
+    stdout: Union[bool, IOBase] = True,
+    stderr: Union[bool, IOBase] = True,
+    **kwargs,
+) -> subprocess.Popen:
+    """Run a command without blocking the calling thread.
+
+    *DO NOT USE THIS FUNCTION* Instead, call subprocess.run or
```

Comment:
I remember this function was developed to help novice python users (in our early days) to easily execute and interact with RG's code to pass arguments. We are probably mature enough to mandate a proper way.

Can you add some risk involved?
Also an example of proper use of subprocess.run will be nice

Reply:
There isn't a whole lot of risk per-se. The main reason not to use it in my opinion is that it provides no real advantage `subprocess.check_call`. I mean the difference between

```python
>>> exe('ls')
# Some popen object
``` 
and 
```python
>>> subprocess.check_output(['ls'])
"file1.py\nfile2.py\n..."
```
is pretty neglible. Arguably the second is more noob friendly because it returns the stdout immediately and this is usually what they want.

I suppose there would be a risk that you might leak the popen object if you forget to wrap it in a context manager, which you do not need to care about for `run`, `check_call`, and friends.